Modules/itertoolsmodule.c

```diff
@@ -3802,6 +3802,75 @@ static PyType_Spec ziplongest_spec = {
 };
 
 
+/* ilen function *************************************************************/
+
+/*[clinic input]
+itertools.ilen
+
+    iterable: object
+    /
+
+Equivalent to len(list(iterable)).
+[clinic start generated code]*/
+
+static PyObject *
+itertools_ilen(PyObject *module, PyObject *iterable)
+/*[clinic end generated code: output=99a1c852a6d09dd4 input=cf51dec933b693da]*/
+{
+    PyObject *it, *item;
+    PyObject *(*iternext)(PyObject *);
+    PyObject *long_cnt = NULL;
+    PyObject *new_long_cnt, *temp;
+    Py_ssize_t cnt;
+
+    it = PyObject_GetIter(iterable);
+    if (it == NULL) {
+        return NULL;
+    }
+
+    cnt = 0;
+    iternext = *Py_TYPE(it)->tp_iternext;
```

Comment:
It might be helpful to possibly use the `__length_hint__` if available. For instance, if the latter gives you something way below the SSIZE_T_MAX, you can do a while loop without the `if` check and then another while loop with the `if`. More precisely, you can do a fast while loop (a while loop without `if cnt == PY_SSIZE_T_MAX`) up to `min(__length_hint__, PY_SSIZE_T_MAX)`, check if there are remaining items and if so, do the slow while loop.

NB: In the slow while loop, you can split it into two as well where `long_cnt == NULL` and when it's been set once but since this condition only holds for a single `cnt` it does not really matter.

Reply:
Thanks for some food for thought.

The issue is... How would the fast loop look like?
```C
for (i=0; i<= PY_SSIZE_T_MAX, i++)
```
does the same `if` check implicitly.

Also, splitting into 2 loops where `long_cnt == NULL` and the "rest" will not result in much savings as the body of `if (cnt == PY_SSIZE_T_MAX) {` will be accessed once in 2147483647 on 32bit machine.

Also, comparison of timing if I replace PY_SSIZE_T_MAX with "1" so that it is accessed on every loop:
```
a = range(100_000)
%timeit ilen(a)                # 1.45 ms
%timeit ilen_pylong(a)   # 3.72 ms
```

Reply:
Hmm, wouldn't you need a for loop then?

```c
for (cnt = 0; cnt < length_hint && (item = iternext(it)) != NULL; cnt++) {
    Py_DECREF(item);
}

Is that even faster?

Reply:
Checked it. Same runtime, no improvement at all.

Reply:
> The issue is... How would the fast loop look like?

You can do a `while` and decrement `length_hint` in the loop condition itself and check for whether its `>0` or not. If at the end of the loop you hit `item != NULL` then you know that you've consumed `length_hint` items. If not, then you know that `length_hint` was an underestimation so you'll need to consume more, e.g.:

```C
rem = length_hint;
while ((item = iternext(it)) != NULL && --rem) {
	Py_DECREF(item);
}
// now either item == NULL or rem == 0
if (item != NULL) {
	// still things to process, do the slow loop
} else {
	cnt = length_hint - rem;
}
```

> Also, splitting into 2 loops where long_cnt == NULL and the "rest" will not result in much savings as the body of if (cnt == PY_SSIZE_T_MAX) { will be accessed once in 2147483647 on 32bit machine.

Yes, that's why I said it does not matter (I think I should not have mentioned at all :laughing:)

> Is that even faster?

This is something I am not sure about, hence the suggestion. I did not check the assembly myself for this one but I think it could be faster. At least, the timings

Reply:
Ah I took too long for my answer. So using `__length_hint__` does not help at all, well sometimes I get surprises (now that I think about, we still have two ifs so it's maybe not really a surprise).

Reply:
Hm, your code would be faster for sure, but there is an issue - what if `length_hint` is lying by a lot? There is a risk of hitting bound on a negative side.

Hmmm...

Reply:
- If it's underestimating, you'll reach the length_hint = 0 and exit the while loop so all's fine.
- If it's overestimating, you'll reach the `item == NULL` before and exit the while before being able to make `--length_hint` so you'll stay `>= 0`.

Note that you need to replace `length_hint` by `min(length_hint, PY_SSIZE_T_MAX)` so that you know that you'll not overflow on a negative by a lot. By the way, I did not checked my algorithm that I wrote (like I might have an 1-offset).

Reply:
> * If it's underestimating, you'll reach the length_hint = 0 and exit the while loop so all's fine.

But there is no condition to exit at `0`, if we add one, then we loose performance gain.

Reply:
```C
while ((item = iternext(it)) != NULL && --rem) {
	Py_DECREF(item);
}
```

The condition is ``--rem``.

Reply:
Then maybe no need for `length_hint` at all. Can just do `PY_SSIZE_T_MAX`.

Reply:
Mmh, maybe yes. Though you'll need to handle the remainder differently then. But again, reversed loops are not always faster than forward loops, but if we strive for the fastest possible code (especially motivated by the purpose of `ilen`), then we should consider different alternatives.

Reply:
I can not observe any noticeable improvement. I think gain here is too small to sacrifice simplicity.

Reply:
You can mark this discussion as resolved (sorry for a false hope).